Encryption.md

```diff
@@ -133,48 +159,144 @@ Algorithms section. The length of the encryption buffer (a 4-byte little endian)
+  
+  /** Retrieval metadata of the column-specific key **/
+  2: optional binary column_key_metadata
+}
+
+union ColumnCryptoMetaData {
+  1: EncryptionWithFooterKey ENCRYPTION_WITH_FOOTER_KEY
+  2: EncryptionWithColumnKey ENCRYPTION_WITH_COLUMN_KEY
+}
+
+struct ColumnChunk {
+...
+  /** Crypto metadata of encrypted columns **/
+  8: optional ColumnCryptoMetaData crypto_meta_data
+}
+```
+
+The `path_in_schema` field in  `EncryptionWithColumnKey` structure is needed because the 
+column name (path) is kept inside the encrypted `ColumnMetaData`. The column name can be used 
+for explicit passing of a decryption key for this column, without calling a key retrieval callback. 
+
+### Encryption key metadata
+A wide variety of services and tools for management of encryption keys exist in the industry 
+today. Public clouds offer different key management services (KMS), and organizational IT 
+systems either build proprietary key managers in-house or adopt open source tools for 
+on-premises deployment. Besides the diversity of management tools, there are many ways 
+to generate and handle the keys themselves (generate Data keys inside KMS – or locally upon 
+data encryption; use Data keys only, or use Master keys to encrypt/wrap the Data keys; store 
+the wrapped key material inside the data file, or at a separate location; etc). There is also 
+a large variety of authorization and certification methods, required to control the access to 
+encryption keys.
+
+Parquet is not limited to a single KMS, key generation/wrapping method, or authorization 
+service. Instead, Parquet provides a developer with a simple interface (key_metadata byte 
+array, and a key retrieval callback) that can be utilized for implementation of any key 
+management scheme. For example, the key_metadata can keep a serialized 
+
+   * String ID of a Data encryption key (for key retrieval from KMS).
+   * Wrapped Data key, and string ID of a Master key.  The Data key is generated randomly and 
+   wrapped either remotely in a KMS, or locally after retrieving the Master key from KMS. 
+   Wrapping format depends on the KMS and can be a JSON string, or a base64 encoded byte array 
+   in a case of local wrapping.
+   * Short ID (counter) of a key inside the Parquet data file. The Data key is wrapped with a Master 
+   key using one of the options described above – but the resulting key material is stored separately, 
+   outside the data file, and will be retrieved using the counter and file path.
+   * Any of the three above, plus a string ID of the KMS instance (in case there are many).
+   * Random string (useful for creation of footer keys in organizations not willing to 
+   authenticate and connect all readers to a KMS. The footer key is generated as F(key_metadata), 
+   where F is an algorithm unknown outside organization. These keys are less secure than 
+   KMS-managed keys, but better than using no keys at all in plaintext footers, see a discussion below).
+
```

Comment:
Key metadata can also be empty (not set) - in case the keys are fully managed by the caller code, and passed explicitly to Parquet for the footer and each encrypted column. I'll add this to the text.